### PR TITLE
Add web page scraping feature implemented by Playwright/Selenium

### PR DESCRIPTION
**Pull request by shenchucheng**
- Support Selenium apis
- Support Playwright apis
- optional dependencies using extras_require: web_browser, playwright, selenium
- tested on both Windows and Ubuntu systems

### CODE REVIEW

**Comment by geekan** [comment]
LGTM

File: metagpt/tools/web_browser_engine.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    async def run(self, url: str, *urls: str) -> str | list[str]:
+        page = await self.run_func(url, *urls)
+        if isinstance(page, str):
+            return get_page_content(page)
+        return [get_page_content(i) for i in page]
+
+
+def get_page_content(page: str):
+    soup = BeautifulSoup(page, "html.parser")
+    return "\n".join(i.text.strip() for i in soup.find_all(["h1", "h2", "h3", "h4", "h5", "p", "pre"]))
```
Comment (better629) [inline metagpt/tools/web_browser_engine.py:53]: It seems that can only get the content only with the html page meets the templates. What if the user want to parse html page with own regex or templates?

File: metagpt/tools/web_browser_engine.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    def __init__(
+        self,
+        engine: WebBrowserEngineType | None = None,
+        run_func: Callable[..., Coroutine[Any, Any, str | list[str]]] | None = None,
+    ):
+        self.config = Config()
+        engine = engine or self.config.web_browser_engine
+
+        if engine == WebBrowserEngineType.PLAYWRIGHT:
+            web_browser_engine = importlib.import_module("metagpt.tools.web_browser_engine_playwright")
```
Comment (better629) [inline metagpt/tools/web_browser_engine.py:24]: the value name ends with `_engine` seems not well, actually it's a module.

File: metagpt/tools/web_browser_engine_selenium.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+from selenium.webdriver.common.by import By
+from selenium.webdriver.support import expected_conditions as EC
+from selenium.webdriver.support.wait import WebDriverWait
+from concurrent import futures
+
+
+class SeleniumWrapper:
+    """Wrapper around Selenium.
+
+    To use this module, you should have the ``selenium`` Python package installed and ensure
```
Comment (better629) [inline metagpt/tools/web_browser_engine_selenium.py:20]: How about doing a selenium packages check before running

File: requirements.txt [hunk 4]
```diff
@@ -29,4 +29,7 @@ tenacity==8.2.2
 tiktoken==0.3.3
 tqdm==4.64.0
 #unstructured[local-inference]
+# playwright
```
Comment (better629) [inline requirements.txt:32]: why comment it ?

File: metagpt/config.py [hunk 5]
```diff
@@ -32,35 +33,42 @@ class Config(metaclass=Singleton):
...
-
-        self.claude_api_key = self._get('Anthropic_API_KEY')
-
-        self.serpapi_api_key = self._get('SERPAPI_API_KEY')
-        self.serper_api_key = self._get('SERPER_API_KEY')
-        self.google_api_key = self._get('GOOGLE_API_KEY')
-        self.google_cse_id = self._get('GOOGLE_CSE_ID')
-        self.search_engine = self._get('SEARCH_ENGINE', SearchEngineType.SERPAPI_GOOGLE)
-        self.max_budget = self._get('MAX_BUDGET', 10.0)
+        self.openai_api_base = self._get("OPENAI_API_BASE")
```
[the remark this answers is not in the document]
Comment (shenchucheng) [reply metagpt/config.py:49]: The code was just formatted by black, and black uses double quotes as the default quotation mark. Do I need to change double quotes to single quotes?

File: metagpt/tools/web_browser_engine.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+from bs4 import BeautifulSoup
+
+
+class WebBrowserEngine:
+    def __init__(
+        self,
+        engine: WebBrowserEngineType | None = None,
+        run_func: Callable[..., Coroutine[Any, Any, str | list[str]]] | None = None,
+    ):
+        self.config = Config()
```
[the remark this answers is not in the document]
Comment (shenchucheng) [reply metagpt/tools/web_browser_engine.py:20]: Oh, I will use `CONFIG` instead. By the way, the line `self.config = Config()` appears in many files.

File: metagpt/config.py [hunk 7]
```diff
@@ -32,35 +33,42 @@ class Config(metaclass=Singleton):
...
+        self.openai_api_model = self._get("OPENAI_API_MODEL", "gpt-4")
+        self.max_tokens_rsp = self._get("MAX_TOKENS", 2048)
+        self.deployment_id = self._get("DEPLOYMENT_ID")
+
+        self.serpapi_api_key = self._get("SERPAPI_API_KEY")
+        self.serper_api_key = self._get("SERPER_API_KEY")
+        self.google_api_key = self._get("GOOGLE_API_KEY")
+        self.google_cse_id = self._get("GOOGLE_CSE_ID")
+        self.search_engine = self._get("SEARCH_ENGINE", SearchEngineType.SERPAPI_GOOGLE)
+        self.web_browser_engine = self._get("WEB_BROWSER_ENGINE", WebBrowserEngineType.PLAYWRIGHT)
```
[the remark this answers is not in the document]
Comment (shenchucheng) [reply metagpt/config.py:68]: `WEB_BROWSER_ENGINE` is an enumeration type, and it's impossible to obtain the correct value from the configuration/environment variables. The same issue exists in `SEARCH_ENGINE`. To fix this, the code could be changed to `self.web_browser_engine = WebBrowserEngineType(self._get("WEB_BROWSER_ENGINE", "playwright"))`.

File: metagpt/tools/web_browser_engine.py (hunk 1 above)
Comment (shenchucheng) [reply metagpt/tools/web_browser_engine.py:54]: [add parse_func for WebBrowserEngin to support custom parsing](https://github.com/geekan/MetaGPT/pull/69/commits/0700cd2e62477227034017ea3fd8bc65f98446cb)

File: metagpt/tools/web_browser_engine.py (hunk 2 above)
Comment (shenchucheng) [reply metagpt/tools/web_browser_engine.py]: [remove inappropriate variable names used in the definition of the class WebBrowserEngine](https://github.com/geekan/MetaGPT/pull/69/commits/a6f13b693847298123f3b84deee571b8b17d880f)

File: metagpt/tools/web_browser_engine_selenium.py (hunk 3 above)
Comment (shenchucheng) [reply metagpt/tools/web_browser_engine_selenium.py]: 1. Run the following command: pip install metagpt[selenium].
2. Make sure you have a compatible web browser installed and the appropriate WebDriver set up for that browser before running. For example, if you have Mozilla Firefox installed on your computer, you can set the configuration SELENIUM_BROWSER_TYPE to firefox. After that, you can scrape web pages using the Selenium WebBrowserEngine.

File: metagpt/config.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (better629) [reply metagpt/config.py:49]: do it later

**Comment by better629** [comment]
LGTM

**Pull request merged**